### PR TITLE
chore: Update to mypy 1.4.1

### PR DESCRIPTION
**Pull request by tayritenour**
## Description

[MLG-839]

mypy==1.4.1 is the highest version we can use without also updating python>=3.8 (currently at 3.7.x)

## Test Plan
CircleCI passes. No need for release party validation.

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

## Ticket

[MLG-839]: https://hpe-aiatscale.atlassian.net/browse/MLG-839?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by hamidzr — APPROVED** [review]
lgtm but probably needs a review from @determined-ai/ml-sys  as well because of the harness changes

do you know where we ware with our plans to stop supporting py 3.7?

File: e2e_tests/tests/cluster/test_webhooks.py [hunk 1]
```diff
@@ -22,7 +22,7 @@ class WebhookRequestHandler(SimpleHTTPRequestHandler):
     def do_POST(self) -> None:
         global request_to_webhook_endpoint
         global keep_server_running
-        content_length = int(self.headers.get("content-length"))
+        content_length = int(self.headers.get("content-length", 0))
```
Comment (MikhailKardash) [inline e2e_tests/tests/cluster/test_webhooks.py:25]: Did this fail before when the get request failed? Looks like default used to be `None`. Not sure if this change will cover up potential regressions. Maybe have someone from backend review?

File: harness/determined/core/_log_shipper.py [hunk 2]
```diff
@@ -89,7 +89,7 @@ class _ShutdownMessage:
 class _LogSender(threading.Thread):
     def __init__(self, session: api.Session, logs_metadata: Dict) -> None:
         self._queue = queue.Queue(maxsize=SHIP_QUEUE_MAX_SIZE)  # type: queue.Queue[_QueueElement]
-        self._logs = collections.deque()  # type: collections.deque[str]
+        self._logs = collections.deque()  # type: Deque[str]
```
Comment (MikhailKardash) [inline harness/determined/core/_log_shipper.py:92]: `typing.Deque` is technically "deprecated" https://docs.python.org/3/library/typing.html

File: harness/determined/pytorch/_lr_scheduler.py [hunk 3]
```diff
@@ -75,11 +75,11 @@ def __init__(
     def step(self, *args: Any, **kwargs: Any) -> None:
         self._scheduler.step(*args, **kwargs)
 
-    def get_last_lr(self) -> List:
-        return self._scheduler.get_last_lr()
+    def get_last_lr(self) -> List[float]:
+        return list(self._scheduler.get_last_lr())
```
Comment (MikhailKardash) [inline harness/determined/pytorch/_lr_scheduler.py:79]: I think the `list()` is not intuitive here (It only serves the purpose of having the linter not freak out. Probably better to  `type: ignore` to avoid confusion on the return type of `get_last_lr()`.

File: harness/determined/pytorch/_lr_scheduler.py [hunk 4]
```diff
@@ -75,11 +75,11 @@ def __init__(
...
 
     def load_state_dict(self, state_dict: Dict[Any, Any]) -> None:
         self._scheduler.load_state_dict(state_dict)
 
     def state_dict(self) -> Dict[Any, Any]:
-        return self._scheduler.state_dict()
+        return dict(self._scheduler.state_dict())
```
Comment (MikhailKardash) [inline harness/determined/pytorch/_lr_scheduler.py:85]: Same as last comment, probably better to `type: ignore`. `state_dict()` already returns a dict, PyTorch just doesn't like adding type annotations.

**Review by MikhailKardash — COMMENTED** [review]
I think using `type: ignore` when outputs of external libraries ( e.g. `torch`, `numpy` ) are not strongly typed is less confusing to devs and results in more readable code.

File: e2e_tests/tests/cluster/test_webhooks.py (hunk 1 above)
Comment (MikhailKardash) [reply e2e_tests/tests/cluster/test_webhooks.py:25]: Oh, I didn't notice that backend already signed off, it's probably fine then. Feel free to mark this as resolved. :)

File: model_hub/model_hub/utils.py [hunk 5]
```diff
@@ -28,15 +28,15 @@ def expand_like(arrays: List[np.ndarray], fill: float = -100) -> np.ndarray:
...
+        return np.ndarray(np.concatenate(arrays))
     full_shape[0] = sum(a.shape[0] for a in arrays)
     full_shape[1] = max(a.shape[1] for a in arrays)
     result = np.full(full_shape, fill)
     row_offset = 0
     for a in arrays:
         result[row_offset : row_offset + a.shape[0], : a.shape[1]] = a
         row_offset += a.shape[0]
-    return result
+    return result  # type: ignore
```
Comment (MikhailKardash) [inline model_hub/model_hub/utils.py:39]: I'd decide on one of `type: ignore` or `np.ndarray(result)`. You're using both throughout this function.

File: harness/determined/pytorch/_lr_scheduler.py (hunk 4 above)
Comment (wes-turner) [reply harness/determined/pytorch/_lr_scheduler.py:85]: In order of preference, I suggest:
1. adding annotations
2. disabling [the specific type error](https://mypy.readthedocs.io/en/stable/error_codes.html)
3. disabling all type errors
